### PR TITLE
some test and scoping fixes

### PR DESCRIPTION
**Pull request by goto40**
I hope I understood right, that I rebase against this branch...

## Code review checklist

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [ ] Changelog(s) has/have been updated, as needed (see `CHANGELOG.md`, no need
      to update for typo fixes and such).

[commit messages]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by goto40** [comment]
I still have the effect with the missing exception (still investigating).

**Comment by goto40** [comment]
I finally identified the issue with the user class "Model".
 * There were several issues with `_getattr` etc. (including a missing `_delattr`) - I fear it is not the end (but we will reach this point while adding more tests).
 * The most important point was, that the root model needs _tx_filename (and some more aspects maybe when resolving multifile models... - not yet fully analyzed)
 * Moreover `attr` strips leading `_`for parameters when calling `__init__`... For this point we need a good solution to flexibly adapt the parameters... (see **TODO** in  model.py).

@markusschmaus @igordejanovic open for comments on this...

**Comment by markusschmaus** [comment]
I don't see the problem with the skipped underscore, I am using such attributes in some of my classes in my private project and it works fine. I specify the attribute in the grammar without the underscore and the only impact is that this attribute is no longer accessible once the object gets initialized, but this happens only at the end anyway.

However I do think it would be good to allow object processors to change the dictionary passed to `__init__` rather than the initialized object itself. For frozen objects it would be too late otherwise, and this could also help with the in some use cases regarding the underscore.

I was thinking that we might want to add a flag to `pre_init` to `register_obj_processors`, which defaults to `False` and then

File: textx/model.py [hunk 1]
```diff
@@ -658,8 +658,13 @@ def call_obj_processors(metamodel, model_obj,
 
                 for m in models:
                     # TODO: what does this check?
-                    assert not m._tx_reference_resolver.parser._inst_stack
+                    assert not _getattr(m, "_tx_reference_resolver").parser._inst_stack
```
Comment (igordejanovic) [inline textx/model.py]: @goto40 I saw this TODO comment in the code. This line make sure that we have processed all parsed instances kept on `_inst_stack` stack. If at this point there is any instance on the stack we did something wrong.

**Comment by igordejanovic** [comment]
@goto40 
> I hope I understood right, that I rebase against this branch...

Yes, you did. BTW, you can drop commit range 
```
ac50c3d | * Style change for ci test
bc85614 | * quick fixed some individual problems
836dc09 | * Use additional set of ids to avoid relying on __eq__ of user class
ae237ef | * Reverse order of object initialization, so children get initialized before their parent
6c7f394 | * Move initialization before object processors
ccbeacd | * Store attributes of user classes outside object and postpone initialization
```

from this branch as those changes are already included in the `feature/postponed-userobj-init`

**Comment by igordejanovic** [comment]
> I finally identified the issue with the user class "Model".

Great work!

> There were several issues with _getattr etc. (including a missing _delattr) - I fear it is not the end (but we will reach this point while adding more tests).

Yeah, I don't feel good about this API change. We should try to find another, backward compatible solution if possible.

> The most important point was, that the root model needs _tx_filename (and some more aspects maybe when resolving multifile models... - not yet fully analyzed)

Yes, the root of the model is special in that regard that there we keep various meta-data. I would go for a simple solution, prohibiting for a root of the model to be immutable/__slots__ class. Trying to fix that is a lot of work and introducing a lot complexity in the code for supporting a use-case that in my view will be very rare.

> Moreover attr strips leading _for parameters when calling __init__... For this point we need a good solution to flexibly adapt the parameters... (see TODO in model.py).

I don't think that should be our issue if that is the way `attr` works. The user of `attr` should be aware and adapt accordingly. Also, like in the case above, we should always strive for simplicity and never try to find a solution for some rare edge cases in advance.

**Pull request closed without merging**